src/data/padics/padic_norm.lean

```diff
@@ -298,6 +317,145 @@ theorem min_le_padic_val_rat_add {q r : ℚ}
+      padic_val_rat.div p (nat.cast_ne_zero.mpr b_split) (nat.cast_ne_zero.mpr (nat.prime.ne_zero p_prime)),
+    rw padic_val_rat.padic_val_rat_self (nat.prime.one_lt p_prime) at split_frac,
+    have r : 1 ≤ padic_val_nat p b := one_le_padic_val_nat_of_dvd b_split dvd,
+    exact_mod_cast split_frac, }
+end
+
+/--
+If a prime doesn't appear in `n`, `padic_val_nat p n` is `0`.
+-/
+lemma padic_val_nat_of_not_dvd {p : ℕ} [fact p.prime] {n : ℕ} (not_dvd : ¬(p ∣ n)) :
+  padic_val_nat p n = 0 :=
+begin
+  by_cases hn : n = 0,
+  { subst hn, simp at not_dvd, trivial, },
+  { rw padic_val_nat_def hn,
+    exact (@multiplicity.unique' _ _ _ p n 0 (by simp) (by simpa using not_dvd)).symm,
+    assumption, },
+end
+
+lemma padic_val_nat_primes {p q : ℕ} [p_prime : fact p.prime] [q_prime : fact q.prime] (neq : p ≠ q) :
+  padic_val_nat p q = 0 :=
+@padic_val_nat_of_not_dvd p p_prime q $ (not_congr (iff.symm (prime_dvd_prime_iff_eq p_prime q_prime))).mp neq
+
+/--
+Dividing out by a prime which is not a factor doesn't change the `padic_val_nat`.
+-/
+-- The strange naming order here, with the two-tick version appearing before the one-tick version,
+-- is because this is a special case of the one-tick version which is nevertheless of interest in
+-- its own right.
+lemma padic_val_nat_div'' {p : ℕ} [p_prime : fact p.prime] {q : ℕ} [q_prime : fact q.prime] (neq : p ≠ q)
+  {b : ℕ} (dvd : q ∣ b) : padic_val_nat p (b / q) = padic_val_nat p b :=
+begin
+  by_cases b_split : b = 0,
+  { rw [b_split, nat.zero_div], },
+  { have q_nonzero : q ≠ 0, by { intro q_zero, simpa [q_zero, zero_dvd_iff] using dvd, },
+    have e : padic_val_rat p (b / q) = padic_val_rat p b - padic_val_rat p q :=
+      padic_val_rat.div p (nat.cast_ne_zero.mpr b_split) (nat.cast_ne_zero.mpr q_nonzero),
+    norm_cast at e,
+    rw [padic_val_nat_primes neq, int.coe_nat_zero, sub_zero] at e,
+    exact int.coe_nat_inj e }
+end
+
+lemma padic_val_nat_div' {p : ℕ} [p_prime : fact p.prime] :
+  ∀ {m : ℕ} (cpm : coprime p m) {b : ℕ} (dvd : m ∣ b), padic_val_nat p (b / m) = padic_val_nat p b
+| 0 := λ cpm b dvd, by { rw zero_dvd_iff at dvd, rw [dvd, nat.zero_div], }
+| (n + 1) :=
+  λ cpm b dvd,
+  begin
+    rcases dvd with ⟨c, rfl⟩,
+    rw [mul_div_right c (nat.succ_pos _)],by_cases hc : c = 0,
+    { rw [hc, mul_zero] },
+    { apply int.coe_nat_inj,
+      simp only [padic_val_rat_of_nat, nat.cast_mul],
+      rw [padic_val_rat.mul]; norm_cast, -- we really need padic_val_nat_mul
+      { suffices : ¬ p ∣ (n+1),
+        { rw [padic_val_nat_of_not_dvd this, zero_add] },
+        contrapose! cpm,
+        exact p_prime.dvd_iff_not_coprime.mp cpm },
+      { exact nat.succ_ne_zero _ },
+      { exact hc } },
+  end
+
+open_locale big_operators
+open_locale classical
+
+-- move this
+lemma finset.prod_multiset_count {M : Type*} [comm_monoid M] (s : multiset M) :
+  s.prod = ∏ m in s.to_finset, m ^ (s.count m) :=
+sorry
```

Comment:
See https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/prod_multiset_count/near/202774761 for proofs and discussion.